recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,401 @@
+          subscriptionid: 1234,
+            secrets: {
+              source: secretStoreAz.id
+            }
+          }
+        },
+        {
+          name: 'azurerm',
+          properties: {
+            subscriptionid: 1234,
+            alias: 'az-paymentservice',
+            secrets: {
+              source: secretStoreAzPayment.id
+            }
+          }
+        }]
+...       
+```
+Configuration for providers as described in this document will take precedence over provider credentials stored in UCP (currently these include azurerm, aws, kubernetes providers). So, for eg. In the scenario where credentials for Azure are saved with UCP during Radius install and a Terraform recipe created by an operator declares 'azurerm' under the the *required_providers* block; If there exists a provider configuration under the *providers* block under *recipeConfig*, these would take precedence and used to build the Terraform configuration file instead of Azure credentials stored in UCP. 
+
+
+Example :
+``` diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
+    recipeConfig: {
+      terraform: {
+        authentication:{
+          ...        
+        }
++        providers: [
++          {
++            name: 'azurerm',
++            properties: {
++              subscriptionid: 1234,
++              secrets: {
++                source: secretStoreAz.id
++              }
++            }
++          },
++          {
++            name: 'azurerm',
++            properties: {
++              subscriptionid: 1234,
++              alias: 'az-paymentservice',
++              secrets: {
++                source: secretStoreAzPayment.id
++              }
++            }
++          },
++          {
++            name: 'gcp',
++            properties: {
++              project: 1234,
++              regions: ['us-east1', 'us-west1']
++            }
++          },
++          {
++            name: 'oraclepass',
++            properties: {
++              database_endpoint: "...",
++              java_endpoint: "...",
++              mysql_endpoint: "..."
++            }
++          }
++        ]
++      }
++      env: {
++        key: value
++        envSecrets: [                    // Individual Secrets from SecretStore
++          {
++             name: 'SPECIAL_LEVEL_KEY'
++             valueFrom: {
++               source: secretStoreConfig.id
++               name: 'special-config'
++               key: 'special.level'
++             }
++          },
++          {
++             name: 'LOG_LEVEL'
++             valueFrom: {
++               source: secretStoreConfig.id
++               name: 'special-config'
++               key: 'special.log'
++             }
++          }
++        ]
++        envSecretsFrom: {               // All Secrets from Secret Store
```

Comment:
Is there a way to merge `envSecretsFrom` and `envSecrets`? Like JavaScript does with the spread operator. Not that important but having two keys (`envSecretsFrom` and `envSecrets`) that look a lot like each other could create confusion.

Reply:
fyi.. envSecretsFrom was removed from the datamodel